packages/data/src/registry.js

```diff
@@ -75,6 +75,26 @@ export function createRegistry( storeConfigs = {}, parent = null ) {
 		return emitter.subscribe( listener );
 	};
 
+	/**
+	 * Subscribe to a state change. Starts watch when the test function returns
+	 * true, then calls the listener when test function returns false.
+	 *
+	 * @param {Function} listener   Listener function
+	 * @param {Function} changeTest Test function to call.
+	 */
+	function subscribeOnChange( listener, changeTest ) {
+		let watching = false;
+		const unsubscribe = subscribe( () => {
+			if ( ! watching && changeTest() ) {
+				watching = true;
+			}
+			if ( watching && ! changeTest() ) {
+				unsubscribe();
+				listener();
+			}
+		} );
+	}
```

Comment:
In the use case that inspired this addition, i.e., observing how post is being saved, the `changeTest` function needs to select something from a data store. But to do that, it would need a `registry` parameter, but it doesn't get one. It can ever work with the global default registry, not with a custom one.

Instead of adding a new method to `registry`, we could implement this functionality also with a helper function that merely combines existing `registry` APIs:
```js
function waitForTransition( registry, test, callback ) {
  registry.subscribe( () => {
    ...
  } );
}
```

It took me a while to figure out what the function does, before I realized that it looks for a transition of the `test()` value from `true` to `false`. It should be documented in these terms.

A unit test for the new functionality should be added.

Reply:
One more thing: the helper could also be a function that returns a Promise. The current `listener` callback is ever executed at most once. Good case for a Promise that resolves when a condition is satisfied.

Reply:
Thanks for the review and suggestions @jsnajdr. I'll try shifting this to a helper function, would it still live in the same file? I'll also work to  improve the docs and add tests.

Reply:
> would it still live in the same file?

It would certainly be a top-level named export of the `@wordpress/data` package. Regarding the file to live in, I think the new helper would be similar to the existing `createRegistrySelector` and `createRegistryControl` functions, which are also self-contained functions that don't depend on anything else than its own parameters.

The `factory.js` file could be renamed to `helpers.js` or `utils.js` and start to host the new helper, too. There is nothing particularly factory-ish about the existing helpers.

Reply:
@jsnajdr 
I'm thinking about giving this a try, but still got some questions (because the data-package/redux is still sometimes difficult to understand for me 😅):

`One more thing: the helper could also be a function that returns a Promise. The current listener callback is ever executed at most once. Good case for a Promise that resolves when a condition is satisfied.`

If the function returns a promise, then the caller doesn't get a unsubscribe function and has no way of unsubscribing. Could this a problem?

I'm not completely sure I understand the structure here. Should the helper be a function which is directly used by someone who wants to subscribe to a change value on a store or is this a helper function which generates a function on the store and the `waitForTransition` will be called on a selector/store instance?

Reply:
> If the function returns a promise, then the caller doesn't get a unsubscribe function and has no way of unsubscribing. Could this a problem?

It's a bit of an inefficiency, because you don't have a way to cancel the wait. Unless the helper returns some extended cancellable promise.

The `.then` handler on the returned promise needs to account for the case when the object/component that initiated the way is already gone by the time the wait finishes. If it does only call React `setState`, then it's safe, React will treat the `setState` as noop if the component is already unmounted.

> is this a helper function which generates a function on the store and the `waitForTransition` will be called on a selector/store instance?

The helper certainly needs a `registry` instance to run on. It can be either:
```js
waitForTransition( callback, registry );
```
or
```js
waitForTransition( callback )( registry );
```
where the second form returns a function that is then applied to a given `registry`. And the same function can be applied to different registries.

It's unlikely anyone will use this "multiple registries" scenario, so just the simple `waitForTransition( callback, registry )` is probably better. The `registry` param can be omitted and then it defaults to the default registry.

Reply:
@jsnajdr I have addressed your feedback in the latest commits, can you please review again?

Also, The doc block is no longer generating docs correctly, appreciate any suggestions for how to troubleshoot that issue (I ran `npm run docs:build`)